src/clj/rems/application/model.clj

```diff
@@ -265,49 +263,67 @@
       (assoc ::submitted-answers (::draft-answers application))
       (update :application/first-submitted #(or % (:event/time event)))
       (dissoc ::draft-answers)
-      (assoc :application/state :application.state/submitted)))
+      (assoc :application/state :application.state/submitted)
+      (assoc :application/todo (if (nil? (:application/first-submitted application))
+                                 :new-application
+                                 :resubmitted-application))))
 
 (defmethod event-type-specific-application-view :application.event/returned
-  [application event]
+  [application _event]
   (-> application
       (assoc ::draft-answers (::submitted-answers application)) ; guard against re-submit without saving a new draft
-      (assoc :application/state :application.state/returned)))
+      (assoc :application/state :application.state/returned)
+      (assoc :application/todo nil)))
+
+(defn- update-waiting-for-todo [application]
+  (cond-> (assoc application :application/todo nil)
```

Comment:
Hm, setting todo to nil here seems problematic, as "Action needed" is then rendered empty after another user has done commenting or deciding, even though some kind of action is still needed from the handler.

I'm not quite sure what to do here, though. I think the state of the application is formally close to "New application" -- it's just that the application is not new anymore. One problem here is that strictly speaking "New application" does not indicate an action, so just removing "New" from that does not make that much sense.

Maybe someone else has an idea of what to show to the user in these states.

(I'd also prefer setting an explicit value to todo instead of null value, but that depends a bit on how this turns out...)

Reply:
Another problem with the "empty action" is that it is difficult to determine what was the previous action that has happened (which affects the next action), e.g., has the application just been reviewed by another user and has returned to the handler, or has the application been long enough in the system that it does not count as a new application anymore, etc.

After thinking a bit more, I think one solution for this would be that "Action needed" is changed back to "State", which is, in some senses, a more general term -- it can consist of an action if the action is easy enough to describe, or just more general state in cases where the next action is not easily described or is ambiguous, which makes it more flexible.

After that change, the state could be, e.g., "Reviewed". In short, the state would describe, in decreasing order of priority:

1) The next action if it is unambiguous and easily described,
2) The previous action so that the user can infer what to do next,
3) The plain state of the application (e.g., "New application").

Reply:
Added a "All requests have been responded to" message for those situations.